### PR TITLE
Add an optional `inherit` argument to Module#autoload? (Feature #15777)

### PR DESCRIPTION
**Pull request by byroot**
Ref: https://bugs.ruby-lang.org/issues/15777

Would be useful in Zeitwerk for instance: https://github.com/fxn/zeitwerk/pull/17

Right now we workaround it with `klass.autoload?(:X) && klass.autoload?(:X) != klass.superclass.autoload?(:X)`.

### CODE REVIEW

File: load.c [hunk 1]
```diff
@@ -1142,25 +1142,42 @@ rb_mod_autoload(VALUE mod, VALUE sym, VALUE file)
 
 /*
  *  call-seq:
- *     mod.autoload?(name)   -> String or nil
+ *     mod.autoload?(name, inherit=true)   -> String or nil
  *
  *  Returns _filename_ to be loaded if _name_ is registered as
- *  +autoload+ in the namespace of _mod_.
+ *  +autoload+ in the namespace of _mod_ or one of it's ancestors.
```
Comment (fxn) [inline load.c]: Would be "its" here.

Comment (byroot) [reply load.c]: Thanks. Fixed it.

File: include/ruby/intern.h [hunk 2]
```diff
@@ -952,6 +952,7 @@ VALUE rb_path2class(const char*);
 VALUE rb_class_name(VALUE);
 VALUE rb_autoload_load(VALUE, ID);
 VALUE rb_autoload_p(VALUE, ID);
+VALUE rb_autoload_at_p(VALUE, ID, VALUE);
```
Comment (eregon) [inline include/ruby/intern.h]: Could you move this to `internal.h`?
`include/ruby/intern.h` is basically part of the public C-API (yes, it's confusing).

Comment (byroot) [reply include/ruby/intern.h]: Thanks, that was actually an outstanding question in the ruby-lang thread.

Comment (byroot) [reply include/ruby/intern.h]: Fixed.

File: spec/ruby/core/module/autoload_spec.rb [hunk 3]
```diff
@@ -11,6 +11,18 @@
...
+  it "returns the name of the file that will be autoloaded if an ancestor defined that autoload" do
+    ModuleSpecs::Autoload::Parent.autoload :AnotherAutoload, "another_autoload.rb"
+    ModuleSpecs::Autoload::Child.autoload?(:AnotherAutoload).should == "another_autoload.rb"
+  end
+
+  ruby_version_is "2.7" do
+    it "returns nil if an ancestor defined that autoload but recursion is disabled" do
+      ModuleSpecs::Autoload::Parent.autoload :AnotherAutoload, "another_autoload.rb"
+      ModuleSpecs::Autoload::Child.autoload?(:AnotherAutoload, false).should be_nil
+    end
```
Comment (eregon) [inline spec/ruby/core/module/autoload_spec.rb]: It would be useful to also test inherit=false with a case that doesn't return `nil`, to make the sure the logic is correct too in that case. Could you add a spec for it?

Comment (byroot) [reply spec/ruby/core/module/autoload_spec.rb]: Of course.

Comment (byroot) [reply spec/ruby/core/module/autoload_spec.rb]: Done.
